projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts

```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
-            break;
-          }
-        }
-        if (!found) {
-          memory[count] = this.color[len];
-        }
-        return memory[count];
+      const latMax = Math.max(...lats); const lonMax = Math.max(...lons);
+      coordinates = {
+        type: 'wgs84',
+        square: { latMin, latMax, lonMin, lonMax }
       };
     }
+
+    this.onDataClick(coordinates);
   }
 
-  private getDrawFeatureCollection(): Observable<any> {
-    const featuresFromQueryCoordinates = (coordinates: any): Observable<any[]> => ObservableOf(coordinates
-        ? coordinates.map((coord: any) =>
-            getFeatureFromGeometry(
-              convertLajiEtlCoordinatesToGeometry(coord)
-            )
-        ) : []);
-
-    return featuresFromQueryCoordinates(this.query.coordinates).pipe(map(features => ({
-      type: 'FeatureCollection',
-      features
-    })));
+  private getBoxQuery$(query: WarehouseQueryInterface, page: number): Observable<AggregateQueryResponse> {
+    return this.warehouseService.warehouseQueryAggregateGet(
+      query,
+      this.useFinnishMap
+        ? ['gathering.conversions.ykj10kmCenter.lat', 'gathering.conversions.ykj10kmCenter.lon']
+        : this.getActiveZoomThresholdLevelToAggregateBy(),
+      undefined, this.boxGeometryPageSize, page, true
+    );
   }
 
-  private updateMapData() {
-    if (!this.ready) {
-      return;
-    }
-    const cacheKey = this.getCacheKey(this.query);
-    if (this.currentCacheKey === cacheKey) {
-      return;
-    }
-    this.currentCacheKey = cacheKey;
-    if (this.subDataFetch) {
-      this.subDataFetch.unsubscribe();
-    }
-    this.drawData.featureCollection.features = [];
-    this.drawDataSubscription?.unsubscribe();
-    this.drawDataSubscription = this.getDrawFeatureCollection().subscribe(featureCollection => {
-      this.drawData = {...this.drawData, featureCollection};
-      this.reset = true;
-      this.loading = true;
-      this.showingIndividualPoints = false;
-      this.fetchQueryAndShowOnMap(this.query);
+  private getBoxDataOptions$(query: WarehouseQueryInterface): Observable<ObservationDataOptions> {
+    // do the first query
+    return this.getBoxQuery$(query, 1).pipe(
+      switchMap(firstPage => (
+        forkJoin([
+          of(firstPage),
+          // get remaining pages
+          ...Array.from(new Array(firstPage.lastPage - 1).keys(), (_, i) => i + 2).map(i => this.getBoxQuery$(query, i))
+        ])
+      )),
+      map(allPages => {
+        const firstPage = allPages[0];
+        for (let i = 1; i < allPages.length; i++) {
+          // join features of all pages to the first page
+          firstPage.features.push(...allPages[i].features);
+        }
+        // return the modified first page
+        return {
+          featureCollection: { type: firstPage.type, features: firstPage.features },
+          on: {
+            click: (...args) => this.onAggregateDataClick(args)
+          },
+          total: firstPage.total
+        };
```

Comment:
Collecting all the features with `allPages.reduce()` would be more clear (clear enough so that you wouldn't need the comments).